### DISCUSSION TITLE
An advection simulation problem

### DISCUSSION THREAD

**Discussion by Wolke926** [idaholab/moose#24451] [Q&A Modules: General] [answered]
I'm a new MOOSE user currently trying to use an advection kernel to simulate a simple rigid body motion process for a sphere. The result I expect is that the ball does not change its shape and moves to the right, but the result is that the shape of the sphere changes and it appears to diffuse. Thank you very much for someone who can answer my question.
The source code has been put here.
```
[Mesh]
  [gen_mesh]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 20
    ymin = 0
    ymax = 20
    nx = 40
    ny = 40
  []
[]

[Variables]
  [c]
  []
[]

[ICs]

  [./c]
    type = SpecifiedSmoothCircleIC           
    invalue = 1.0
    outvalue = 0.0
    int_width = 2.0
    x_positions = '4 '
    y_positions = '4 '
    z_positions = '0.0 '
    radii       = '2' 
    variable = c
    [../]
  
[]

[Kernels]
  [advection]
    type = ConservativeAdvection
    variable = c
    velocity = '2 0 0'
    upwinding_type = full
  []

  [time]
    type = TimeDerivative
    variable = c
  []
[]

[BCs]
 [./allow_mass_out]
  type = OutflowBC
  boundary = right
  variable = c
  velocity = '2 0 0'
[../]
   [./injection_left]
    type = InflowBC
    boundary = left
    variable = c
    velocity = '2 0 0'
    inlet_conc = 0
  [../]
[]

[Executioner]
  type = Transient
  petsc_options = '-snes_converged_reason'
  num_steps = 30
  dt = 0.2
  nl_rel_tol = 1e-12
[]

[Outputs]
  exodus = true
[]
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is probably just some numerical diffusivity from the upwinding scheme.
Could you please try to refine the mesh to quantify the effect on the shape of the advected object?

Guillaume

**Reply by Wolke926** [reply]
Thanks for your reply, I just tried to refine the mesh, but the result doesn't change. Do you think it is related to the boundary condition I am using?
<img width="404" alt="Screen Shot 2023-05-23 at 12 47 22 PM" src="https://github.com/idaholab/moose/assets/120539827/69a7df28-0e25-45d1-b63c-0a20c02ff075">

**Reply by GiudGiud** [reply]
maybe from the coarse time scheme then?
can you try bdf2 for the scheme and/or a finer time grid

**Reply by Wolke926** [reply]
I modified the time step to be smaller (dt = 0.05) and added TimeIntegrator bdf2. This time the result has been improved, but there is still a tendency to diffuse. Should I delete the unwinding term to avoid the numerical diffusivity? Is Full upwinding necessary for advection/convection problems? 
![Screen Shot 2023-05-23 at 1 51 50 PM](https://github.com/idaholab/moose/assets/120539827/f1d3174b-6a78-4d12-a91c-eae830625b97)

**Reply by lindsayad** [reply]
If the solution is sufficiently smooth then upwinding is not necessary. TVD schemes are meant to provide upwinding when necessary and fall back to second order central differencing/averaging when the solution is sufficiently smooth

**Reply by lindsayad** [reply]
You could try with `upwinding_type = none` and see whether you experience any oscillations

**Reply by Wolke926** [reply]
I tried, the illustration looks even better, the diffusion almost disappears, although one side of the sphere shape shows some oscillations.

**Reply by lindsayad** [reply]
that's the trade-off. You could try to introduce something like min-mod or van Leer into that class if you wanted to see whether you could get a better blend

**Reply by jiangwen84** [reply]
You can also check some capabilities in Level set module which deals with pure advection equations. 

See https://mooseframework.inl.gov/modules/level_set/example_circle.html

**Reply by ABallisat** [reply]
It sounds like you are trying to advect a sharp interface (the edge of the sphere) which is a notoriously difficult problem. As @jiangwen84 pointed out Level Set is a good bet, and probably your best bet if you just want to advect an interface. You will have to accept the trade off cost of doing some reinitialisation (either PDE based or fast marching, I have done both in MOOSE). You could also look at a Volume of Fluid/Solid approach and then you will need some stabilisation scheme, for example look at KT stabilisation in the Porous Flow module (not sure if that can be used outside of porous flow easily), there are lots of other methods but no golden bullet. Do you need to couple the rigid body motion with other physics?

If you want something that works straight out of the box for pure advection, the level set module is the best place to start.

**Reply by lindsayad** [reply]
We also have pure scalar advection implemented using finite-volume/rDG through the `FVAdvection` finite volume kernel. I am adding more interpolation types in https://github.com/idaholab/moose/pull/24468 (previously only upwind and weighted averaging)

**Reply by jiangwen84** [reply]
@ABallisat did you implement the fast marching method in MOOSE? How is the performance?

**Reply by ABallisat** [reply]
@jiangwen84 yep I wrote the fast marching in an external solve then it gets called through a sub-app. Performance is alright but it is hampered by being essentially a serial algorithm as it only sets one point per loop (just to be clear the implementation is parallel and works across MPI ranks). I didn't get around to doing a parallel algorithm version as I found other issues in the wider model it was a part of. Overall it was good for 2D and narrow band 3D but really suffered for large 3D meshes that I am interested in. I might go back to it later this year if time allows to see if I can get a parallel algorithm version going to improve performance..

**Reply by jiangwen84** [reply]
@ABallisat thanks for your information. Is your fast marching code and moose input file available somewhere? We would like to implement the fast marching method to update the level set function in MOOSE, so your experience can be very useful. thanks.

**Reply by ABallisat** [reply]
@jiangwen84 its currently in the midst of another code but I can move it out into another repo and share it with you, I will try to get around to it this week.

**Reply by jiangwen84** [reply]
@ABallisat sounds great. Thanks!

**Reply by jiangwen84** [reply]
@ABallisat  any updates?

**Reply by ABallisat** [reply]
@jiangwen84 sorry this took me longer to get around to than intended. I have put the code and a demo file in a repo and given you access to it, let me know if you have any issues with it or if I can help explain the implementation

**Reply by jiangwen84** [reply]
@ABallisat I got your codes. Hopefully we can collaborate more on the level set codes in MOOSE. thanks a lot!
